### PR TITLE
refactor correction0/3 algorithm into its own class for clarity

### PR DESCRIPTION
**Pull request by joseph-ireland**

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            double result = components.Select(component => component.Evaluate(distance, x, y)).Sum();
+            result += offset.Interpolate(distance);
+            return SpecialFunctions.Logistic(result) * scale.Interpolate(distance);
+        }
+
+        private L2NormCorrection[] components;
+        private LinearSpline offset, scale;
+
+        private static readonly double[] flow0distances = { 0.0, 1.0, 1.35, 1.7, 2.3, 3 };
+        public static readonly MultiL2NormCorrection FLOW_0 = new MultiL2NormCorrection(
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs:53]: you sure you want to keep 0/3 naming? its kinda hard to read when separated from the movement class, i'd suggest using prev(prev?)/next instead

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                    scale:    new double[] { 0, 0.4, 0.4, 0, 0 }
+                ),
+            });
+
+        private static readonly double[] snap3distances = { 1, 1.5, 2.5, 4, 6, 8 };
+        public static readonly MultiL2NormCorrection SNAP_3 = new MultiL2NormCorrection(
+            distance: snap3distances,
+            offset: new double[] { -2, -2, -3, -5.4, -4.9, -4.9 },
+            scale: new double[] { 1, 1, 1, 1, 1, 1 },
+            components: new[] {
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs:172]: i feel like amount of components isnt immediately obvious here, might want to have some sort of param to show how many components should be there and add checks when creating the class to ensure its correct

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs (hunk 1 above)
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs:53]: also i feel like there should be a brief explanation of flow/snap naming as well

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+using System;
+using System.Linq;
+using MathNet.Numerics;
+using MathNet.Numerics.Interpolation;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Preprocessing
+{
+    public class L2NormCorrection
+    {
+        public L2NormCorrection(double[] distance, double[] x_offset, double[] y_offset, double[] offset, double[] scale)
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs:13]: is x_offset naming allowed in osu? i thought its only for consts

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs (hunk 2 above)
Comment (joseph-ireland) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs:172]: there can be any number of "components", they're just summed up. If you mean the length of each array, I think the InterpolateSorted function should check that, though I'm not sure

**Pull request merged**